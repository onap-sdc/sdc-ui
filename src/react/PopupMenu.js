import React from 'react';
import PropTypes from 'prop-types';
import PopupMenuItem from './PopupMenuItem';

class PopupMenu extends React.Component {
    render() {
        const {
            children = [],
            onMenuItemClick,
            position = {},
            relative
        } = this.props;
        const style = relative ? { left: position.x, top: position.y } : {};

        return (
            <ul
                className={`sdc-menu-list ${relative ? 'relative' : ''}`}
                style={style}>
                {React.Children.toArray(children).map((child, i) =>
                    React.cloneElement(child, {
                        onClick: child.props.onClick || onMenuItemClick,
                        key: i
                    })
                )}
            </ul>
        );
    }
}

PopupMenu.propTypes = {
    relative: PropTypes.bool,
    position: PropTypes.shape({
        x: PropTypes.number,
        y: PropTypes.number
    }),
    onMenuItemClick: PropTypes.func
};

PopupMenu.defaultProps = {
    relative: false
};

export const PopupMenuSeparator = () => <li className="separator" />;

PopupMenu.Separator = PopupMenuSeparator;
PopupMenu.Item = PopupMenuItem;
export default PopupMenu;
