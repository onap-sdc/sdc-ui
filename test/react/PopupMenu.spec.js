import React from 'react';
import PopupMenu from '../../src/react/PopupMenu.js';

import renderer from 'react-test-renderer';
import {mount} from 'enzyme';

class MyPopupForm extends React.Component {
	constructor(props) {
		super(props);
		this.state = {position: {}};
	}

	handleClick(newPos) {
		this.setState({position: newPos});
	}

	getPosition() {
		return this.state.position;
	}

	render() {
		return (
			<form onClick={event => this.handleClick({
				x: event.pageX - event.target.offsetLeft,
				y: event.pageY - event.target.offsetTop
			})}>
				<PopupMenu position={this.state.position} relative onMenuItemClick={() => {}} />
			</form>
		);
	}
}

describe('PopupMenu', () => {
	test('check static menu rendered', () => {
		const menu = renderer.create(<PopupMenu onMenuItemClick={() => {}}/>).toJSON();
		expect(menu).toMatchSnapshot();
	});

	test('check relative menu rendered', () => {
		const menu = renderer.create(<PopupMenu onMenuItemClick={()=> {}} position={{x: 10, y: 10}} relative/>).toJSON();
		expect(menu).toMatchSnapshot();
	});

	test('check separator rendered', () => {
		const separator = renderer.create(<PopupMenu.Separator/>).toJSON();
		expect(separator).toMatchSnapshot();
	});

	test('check position changed', () => {
		const menuForm = mount(<MyPopupForm />);
		const position = menuForm.instance().getPosition();
		expect(position).toEqual({});
		expect(position).toEqual(menuForm.find('ul').props().style);
		menuForm.find('form').simulate('click', {
			target: {offsetLeft: 10, offsetTop: 20},
			pageX: 30, pageY: 50
		});
		const newPosition = menuForm.instance().getPosition();
		expect(newPosition).toEqual({x:20, y:30});
		expect({left: newPosition.x, top: newPosition.y}).toEqual(menuForm.find('ul').props().style);
	});
});