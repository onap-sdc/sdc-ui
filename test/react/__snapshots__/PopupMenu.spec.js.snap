// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupMenu check relative menu rendered 1`] = `
<ul
  className="sdc-menu-list relative"
  style={
    Object {
      "left": 10,
      "top": 10,
    }
  }
/>
`;

exports[`PopupMenu check separator rendered 1`] = `
<li
  className="separator"
/>
`;

exports[`PopupMenu check static menu rendered 1`] = `
<ul
  className="sdc-menu-list "
  style={Object {}}
/>
`;
